src/backend/distributed/relay/relay_event_utility.c

```diff
@@ -622,6 +599,56 @@ RelayEventExtendNames(Node *parseTree, char *schemaName, uint64 shardId)
 }
 
 
+/*
+ * RelayEventExtendConstraintAndIndexNames extends the names of constraints
+ * and indexes in given constraint with the shardId.
+ */
+static void
+RelayEventExtendConstraintAndIndexNames(AlterTableStmt *alterTableStmt,
+										Constraint *constraint,
+										uint64 shardId)
+{
+	char **constraintName = &(constraint->conname);
+	const bool missingOk = false;
+	Oid relationId = RangeVarGetRelid(alterTableStmt->relation,
+									  AccessShareLock,
+									  missingOk);
+
+	if (constraint->indexname)
+	{
+		char **indexName = &(constraint->indexname);
+		AppendShardIdToName(indexName, shardId);
+	}
+
+	/*
+	 * Append shardId to constraint names if
+	 *  - table is not partitioned or
+	 *  - constraint is not a CHECK constraint
+	 *
+	 * We do not want to append shardId to partitioned table shards because
+	 * the names of constraints will be inherited, and the shardId will no
+	 * longer be valid for the child table.
+	 *
+	 * See MergeConstraintsIntoExisting function in Postgres that requires
+	 * inherited check constraints in child tables to have the same name
+	 * with those in parent tables.
+	 */
+	if (!PartitionedTable(relationId) ||
+		constraint->contype != CONSTR_CHECK)
+	{
+		/*
+		 * constraint->conname could be empty in the case of
+		 * ADD {PRIMARY KEY, UNIQUE} USING INDEX.
+		 * In this case, already extended index name will be used by postgres.
+		 */
+		if (constraint->conname != NULL)
+		{
+			AppendShardIdToName(constraintName, shardId);
```

Comment:
are there any hazards around partitioned tables?

Reply:
I wouldn't expect such a problem because we perform the necessary checks via PrepareAlterTableStmtForConstraint beforehand.